pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,527 @@
+// Copyright 2021 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package memberlist
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/golang/groupcache/consistenthash"
+	"github.com/hashicorp/memberlist"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/agent/config"
+	"antrea.io/antrea/pkg/apis/crd/v1alpha2"
+	crdinformers "antrea.io/antrea/pkg/client/informers/externalversions/crd/v1alpha2"
+	crdlister "antrea.io/antrea/pkg/client/listers/crd/v1alpha2"
+	"antrea.io/antrea/pkg/util/k8s"
+)
+
+const (
+	controllerName = "MemberListCluster"
+	// Set resyncPeriod to 0 to disable resyncing.
+	resyncPeriod time.Duration = 0
+	// set default virtual node replicas num of consistent hash
+	defaultVirtualNodeReplicas = 3
+)
+
+type localNodeEventType int
+
+const (
+	localNodeWillJoin localNodeEventType = iota
+	localNodeWillLeave
+)
+
+// default Hash Fn is crc32.ChecksumIEEE
+var defaultHashFn func(data []byte) uint32
+
+type ClusterInterface interface {
+	Run(stopCh <-chan struct{})
+	ShouldSelect(name string) bool
+	AddEventHandler(handler clusterNodeEventHandler)
+}
+
+type clusterNodeEventHandler func(nodeName string, isJoin bool)
+
+type Cluster struct {
+	bindPort int
+	// NodeConfig contains nodeName and ipAddr which would be used
+	NodeConfig *config.NodeConfig
+	// Name of local node. Node name must be unique in the cluster.
+	nodeName string
+
+	mList *memberlist.Memberlist
+	// conHash hold the consistenthash.Map, when a node join cluster, use method Add() to add some keys to the hash.
+	// when a node leave the cluster, the consistenthash.Map should update
+	conHash *consistenthash.Map
+	// 	nodeEventsCh, the Node join/leave events will be notified via it.
+	nodeEventsCh chan memberlist.NodeEvent
+
+	// ClusterNodeEventHandlers hold the fun that should implement when a node join or leave the cluster
+	// for example, when a nwe node join the cluster,
+	// each node should recalculate that if it should still hold the existing egresses owned by itself
+	// in the same way, when a node leave cluster, each node may get some new egresses from the left node
+	ClusterNodeEventHandlers []clusterNodeEventHandler
```

Comment:
I guess this is used in later patch? Maybe adding it later when it's used so we can know whether it's suitable.
And it shouldn't be capitalized as a member variable.

BTW, if you feel having two PRs to support the failover makes it difficult to organize code or doesn't help the review progress, feel free to merge it with https://github.com/antrea-io/antrea/pull/2225. Either works for me. Just want to avoid checking in code that is not used yet.

Reply:
done